OpenRA.Mods.Common/Traits/World/ColorPickerManager.cs

```diff
@@ -179,8 +127,100 @@ Color MakeValid(float hue, float sat, MersenneTwister random, List<(float R, flo
+			return Color.FromAhsv(random.NextFloat(), randomSat, randomVal);
+		}
+
+		#region IColorPickerManagerInfo
+
+		public event Action<Color> OnColorPickerColorUpdate;
+
+		(float hMin, float hMax) IColorPickerManagerInfo.GetHueRange { get; }
+		(float sMin, float sMax) IColorPickerManagerInfo.GetSaturationRange => (HsvSaturationRange[0], HsvSaturationRange[1]);
+		(float vMin, float vMax) IColorPickerManagerInfo.GetValueRange => (HsvValueRange[0], HsvValueRange[1]);
+
+		Color[] IColorPickerManagerInfo.PresetColors => PresetColors;
+
+		Color IColorPickerManagerInfo.RandomPresetColor(MersenneTwister random, IEnumerable<Color> terrainColors, IEnumerable<Color> playerColors)
+		{
+			var terrainLinear = terrainColors.Select(c => c.ToLinear()).ToList();
+			var playerLinear = playerColors.Select(c => c.ToLinear()).ToList();
+
+			foreach (var color in PresetColors.Shuffle(random))
+			{
+				// Color may already be taken
+				var linear = color.ToLinear();
+				if (!TryGetBlockingColor(linear, terrainLinear, out _) && !TryGetBlockingColor(linear, playerLinear, out _))
+					return color;
+			}
+
+			// Fall back to a random non-preset color
+			var randomHue = random.NextFloat();
+			var randomSat = float2.Lerp(HsvSaturationRange[0], HsvSaturationRange[1], random.NextFloat());
+			var randomVal = float2.Lerp(HsvValueRange[0], HsvValueRange[1], random.NextFloat());
+			return MakeValid(randomHue, randomSat, randomVal, random, terrainLinear, playerLinear, null);
 		}
+
+		Color IColorPickerManagerInfo.MakeValid(Color color, MersenneTwister random, IEnumerable<Color> terrainColors, IEnumerable<Color> playerColors, Action<string> onError)
+		{
+			var (_, h, s, v) = color.ToAhsv();
+			return MakeValid(h, s, v, random, terrainColors, playerColors, onError);
+		}
+
+		Color IColorPickerManagerInfo.RandomValidColor(MersenneTwister random, IEnumerable<Color> terrainColors, IEnumerable<Color> playerColors)
+		{
+			var h = random.NextFloat();
+			var s = float2.Lerp(HsvSaturationRange[0], HsvSaturationRange[1], random.NextFloat());
+			var v = float2.Lerp(HsvValueRange[0], HsvValueRange[1], random.NextFloat());
+			return MakeValid(h, s, v, random, terrainColors, playerColors, null);
+		}
+
+		void IColorPickerManagerInfo.ShowColorDropDown(DropDownButtonWidget dropdownButton, Color initialColor, string initialFaction, WorldRenderer worldRenderer, Action<Color> onExit)
+		{
+			dropdownButton.RemovePanel();
+
+			void AddActorPreview(Widget parent)
+			{
+				var preview = parent.GetOrNull<ActorPreviewWidget>("PREVIEW");
+				if (preview == null)
+					return;
+
+				if (initialFaction == null || !FactionPreviewActors.TryGetValue(initialFaction, out var actorType))
+				{
+					if (PreviewActor == null)
+						throw new YamlException($"{nameof(ColorPickerManager)} does not define a preview actor" + initialFaction == null ? "." : $"for faction {initialFaction}.");
```

Comment:
This needs some brackets to evaluate properly:
```suggestion
						throw new YamlException($"{nameof(ColorPickerManager)} does not define a preview actor" + (initialFaction == null ? "." : $"for faction {initialFaction}."));
```